### ISSUE TITLE
Remove payload from DELETE requests

### ISSUE THREAD

**Issue by mastertinner** [cloudfoundry/cloud_controller_ng#609]
## Issue

Some API calls (especially ones that the CLI is using) work with payloads to send data with their DELETE requests. According to [Stackoverflow](https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request) and many other sources, this is to be discouraged. Many frameworks strip the body from DELETE requests.
## Steps to Reproduce

Do a `CF_TRACE=true cf unset-space-role USERNAME ORG SPACE ROLE` and look at the last request.
## Expected result

I would expect CF not to use payloads with DELETE requests but rather use two requests to get the UUID of the user in the above example or use query parameters to send the username. The URL of a DELETE request should always be specific about which entity to delete, which it isn't in the current implementation.
## Current result

In our project, we are using a proxy between the CLI and the Cloud Controller. This proxy uses [Jersey](https://jersey.java.net/) which correctly strips the payload of a DELETE request. This renders the `cf unset-space-role` command not working. Furthermore, the CLI says "OK" after the request has been processed, even though the operation was not successful (no roles were deleted).
## Possible Fix

Use two requests or use query parameters.

**Comment by SocalNick** [comment]
@mastertinner thanks for the bug report, I agree this is less than ideal. I think this feature was added because OrgManagers and SpaceManagers can't actually search for users by username to add to their org / space. Similar on the delete side, it's difficult to discover the GUID for a particular user in order to use the original [Remove Developer from Space endpoint](http://apidocs.cloudfoundry.org/237/spaces/remove_developer_from_the_space.html). @dieucao might be able to provide additional context.

These features can actually be disabled by two feature flags: `set_roles_by_username` and `unset_roles_by_username`. Maybe the CLI can do an alternate workflow when these are disabled?

@dkoper do you know if the CLI can do alternative requests when `unset_roles_by_username` feature flag is disabled and if these requests avoid using a body for DELETE http method?

**Comment by dkoper** [comment]
@SocalNick Not sure I understand the question. I thought that when `unset_roles_by_username` is disabled, the CLI tries to discover the GUID of the user to call the API you linked to - so unless they're an admin, the GUID retrieval fails with a permission error.
What we're doing now is already quite complex, see https://www.pivotaltracker.com/story/show/107375664, and I wonder if the story description is entirely correct.
What alternate workflow would you suggest?

**Comment by SocalNick** [comment]
At least for DELETE, you could find all the users in the org / space by role and then iterate over the returned results for the username.

It would be much better if you could search org / space users by username to get the guid then delete by guid.

OR

We support DELETE with username query param in addition to the body param.

**Comment by dkoper** [comment]
I don't think the latter is an option (https://www.w3.org/Protocols/rfc2616/rfc2616-sec15.html#sec15.1.3).

The following API seems a good candidate?
http://apidocs.cloudfoundry.org/207/spaces/retrieving_the_roles_of_all_users_in_the_space.html

It seems it was introduced in CC 2.25.0, so should be fine.

**Comment by SocalNick** [comment]
@dkoper would be great to try to do the delete by using the API endpoint you mentioned to obtain the user guid. I'll move the tracker story to your backlog.

**Comment by dkoper** [comment]
Thanks. You can track it here:
https://www.pivotaltracker.com/story/show/120515881

**Comment by mastertinner** [comment]
thanks for the quick reply, gentlemen!

**Comment by michaelxupiv** [comment]
Hey @mastertinner,

Thanks for submitting this issue. It looks like this story has been moved to the CLI backlog, so we will close it here for now. Feel free to reopen this issue if you feel like we haven't addressed your concerns.

Thanks,
CAPI Community Pair - @ericpromislow & @michaelxupiv

**Comment by mastertinner** [comment]
Thanks for your reply, @michaelxupiv. For me this issue isn't solved yet. Solving it in the CLI is merely combating symptoms. There might be other clients which are using the same incorrect functionality as the CLI. I would suggest to, as soon as the respective CLI issue is closed, deprecate the faulty endpoints in the CC API. This would mean deprecating:

* [Remove Auditor with the Organization by Username](https://apidocs.cloudfoundry.org/249/organizations/remove_auditor_with_the_organization_by_username.html)
* [Remove Billing Manager with the Organization by Username](https://apidocs.cloudfoundry.org/249/organizations/remove_billing_manager_with_the_organization_by_username.html)
* [Remove Manager with the Organization by Username](https://apidocs.cloudfoundry.org/249/organizations/remove_manager_with_the_organization_by_username.html)
* [Remove User with the Organization by Username](https://apidocs.cloudfoundry.org/249/organizations/remove_user_with_the_organization_by_username.html)
* [Remove Auditor with the Space by Username](https://apidocs.cloudfoundry.org/249/spaces/remove_auditor_with_the_space_by_username.html)
* [Remove Developer with the Space by Username](https://apidocs.cloudfoundry.org/249/spaces/remove_developer_with_the_space_by_username.html)
* [Remove Manager with the Space by Username](https://apidocs.cloudfoundry.org/249/spaces/remove_manager_with_the_space_by_username.html)

So, in summary, I would suggest to reopen this issue and make it dependent on the respective CLI issue.

**Comment by ljfranklin** [comment]
@mastertinner sounds reasonable, we're left [our story](https://www.pivotaltracker.com/story/show/136577507) open and marked it as blocked on the [CLI story](https://www.pivotaltracker.com/n/projects/892938/stories/120515881).

**Comment by zrob** [comment]
Latest CAPI release uses POST endpoints to remove users by username
